src/dtype.rs

```diff
@@ -605,7 +602,7 @@ impl<'py> PyArrayDescrMethods<'py> for Bound<'py, PyArrayDescr> {
         }
     }
 
-    fn names(&self) -> Option<Vec<&str>> {
+    fn names(&self) -> Option<Vec<PyBackedStr>> {
```

Comment:
I wonder if this API isn't too clever after and we should just return `Option<Bound<'_, PyAny>>` instead. The caller can then extract the result into whatever collection type is appropriate (`Vec`, array, using `&str` if they still use GIL refs).

Reply:
Hmm, good point. `Option<Bound<'_, PyAny>>` feels a bit loose on the type side. You'd have to look in the docs to know what you can do with it. At the very least I feel like we should use `Option<Bound<'_, PyIterator>>` to convey that one can iterate it. One step further would be to use `Option<impl Iterator<Item = ...>>` which would allow users to choose the collection (if at all), but we can still provide the item type (in this case probably `PyResult<PyBackedStr>`). What do you think about that?

Reply:
> At the very least I feel like we should use Option<Bound<'_, PyIterator>> to convey that one can iterate it.

This feels nice enough.

> One step further would be to use Option<impl Iterator<Item = ...>> which would allow users to choose the collection (if at all), but we can still provide the item type (in this case probably PyResult<PyBackedStr>).

Fixing the item type is what I would like to avoid so that users can decided whether `PyString` is fine or whether they want to handle `PyBackedStr`.

Personally, I see two design points here:
* Maximum convenience since this should never be performance critical, i.e. `Option<Vec<String>>`.
* Maximum flexibility so downstream code can decided what to with this, i.e. `Option<Bound<'_, PyIterator>>`.

Reply:
It seem like with `Option<Bound<'_, PyIterator>>` it is not possible anymore to extract them as a `Vec<...>` anymore
```
Error: PyErr { type: <class 'TypeError'>, value: TypeError("'tuple_iterator' object cannot be converted to 'Sequence'"), traceback: None }
```
Without that I think the ergonomics of that API are not that great anymore. So if we want to provide maximum flexibility we have to go for `Option<Bound<'_, PyAny>>` I think. The example below shows what could be done in that case.

```rust
let dt = PyArrayDescr::new_bound(py, &[("a", "O"), ("b", "?")])?;

// extract the names by copying them into Rust
let names = dt.names().unwrap().extract::<Vec<String>>()?;
assert_eq!(names, vec!["a".to_owned(), "b".to_owned()]);

// extract the names, but referencing them in the Python heap
let names = dt.names().unwrap().extract::<Vec<PyBackedStr>>()?;
for (name, name_ref) in names.iter().zip(["a", "b"]) {
    assert_eq!(&**name, name_ref)
}

// iterate over the names without allocation
for (value, name_ref) in dt.names().unwrap().iter()?.zip(["a", "b"]) {
    let value = value?;
    let name = value.extract::<&str>()?;
    assert_eq!(name, name_ref);
}
```

But maybe we are making this more complex than it needs to be and we should just go with `Option<Vec<String>>`?

Reply:
> But maybe we are making this more complex than it needs to be and we should just go with Option<Vec<String>>?

Indeed. I do not think this will ever be in the "data path" of an application and hence `Option<Vec<String>>>´ appears as the most reasonable choice to me.